### PR TITLE
Get recursive tasks' statuses with task manager api call

### PR DESCRIPTION
**Pull request by Deexie**
The PR adds an api call allowing to get the statuses of a given
task and all its descendants.

The parent-child tree is traversed in BFS order and the list of
statuses is returned to user.

### CODE REVIEW

**Comment by Deexie** [comment]
@bhalevy review please

File: api/api-doc/task_manager.json [hunk 1]
```diff
@@ -264,6 +264,13 @@
             "progress_completed":{
                "type":"double",
                "description":"The number of units completed so far"
+            },
+            "child_tasks":{
+               "type":"array",
+                "items":{
+                    "type":"string"
+                },
+               "description":"Ids of indirect children of the task"
```
Comment (bhalevy) [inline api/api-doc/task_manager.json:300]: why "indirect"?

File: api/task_manager.cc [hunk 2]
```diff
@@ -35,6 +35,9 @@ struct full_task_status {
     std::string module;
     tasks::task_id parent_id;
     tasks::is_abortable abortable;
+    const tasks::task_manager::foreign_task_vector& child_tasks;
+
+    full_task_status(const tasks::task_manager::foreign_task_vector& child_tasks_) : child_tasks(child_tasks_) {}
```
Comment (bhalevy) [inline api/task_manager.cc:40]: I think it'd be better to copy the task ids here into a vector in full_task_status

File: api/task_manager.cc [hunk 3]
```diff
@@ -81,6 +84,11 @@ tm::task_status make_status(full_task_status status) {
     res.progress_units = status.task_status.progress_units;
     res.progress_total = status.progress.total;
     res.progress_completed = status.progress.completed;
+    std::vector<std::string> ct{status.child_tasks.size()};
+    boost::transform(status.child_tasks, ct.begin(), [] (const auto& child) {
+        return child->id().to_sstring();
+    });
```
Comment (bhalevy) [inline api/task_manager.cc:90]: This conversion should happen earlier and here the id's should be returned as strings in the json result

File: api/task_manager.cc [hunk 4]
```diff
@@ -180,7 +181,51 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+        }));
+        res.push_back(co_await retrieve_status(task));
+
+        // Push children's statuses in BFS order.
+        size_t current = 0;
+        while (current < res.size()) {
+            auto& current_status = res[current];
+
+            for (auto& t : current_status.child_tasks) {
+                res.push_back(co_await retrieve_status(t));
```
Comment (bhalevy) [inline api/task_manager.cc:211]: Since this function yields, it should rely on this vector being valid.
Instead, like I suggested earlier, the task status should keep the children ids and they should be looked up here.

File: repair/repair_task.hh [hunk 5]
```diff
@@ -124,6 +124,9 @@ public:
             streaming::stream_reason reason_,
             abort_source* as,
             bool hints_batchlog_flushed);
+    virtual tasks::is_internal is_internal() const noexcept override {
+        return tasks::is_internal::yes;
+    }
```
Comment (bhalevy) [inline repair/repair_task.hh:129]: please send separately

**Comment by bhalevy** [comment]
> The PR introduces changes to task manager api:
> 
> * extends tasks' list returned with get_tasks with task type,
>   keyspace, table, entity, and sequence number
> * extends status returned with get_task_status and wait_task
>   with a list of children's ids
> * enables retrieving a list of statuses of a task and all
>   its descendants
> 
> Apart from that:
> 
> * task type can be get with type() method (and repair is modified
>   to use it)
> * shard tasks are internal
> * task_manager::module has a virtual destructor

there's a lot going on here.
Pmease send fixes to the existing code bases separaty.
And let"s see if the forward development can be cut into several meaningful independent steps

**Comment by Deexie** [comment]
I split it. New PRs:
- concerning module destructor: #12324
- marking shard tasks as internal: #12325
- accessing task's type with method: #12326

File: api/task_manager.cc (hunk 4 above)
Comment (Deexie) [reply api/task_manager.cc:211]: I extensively focused on minimizing the number of accesses to different shards while unreasonably ignoring all other aspects.

Recursion is done with `current` serving a queue for BFS. Now it would not be so streightforward though.

I can see two approches:

1. In `full_task_status` we keep a vector of task_info (task_id + shard) and look for each task in the module. This is rather problematic, because we could have already removed some of child tasks from module so we won't give user the complete information.

2. We can access information about all descendants even if they were removed from module but we need to simulate a queue somehow (for bfs). To do so we can:
- copy all foreign_ptrs in an actual queue (which means an access to owner shard for each pointer)
- keep a vector of reference wrappers to vectors of children (something similar to what I tried to achieve in first version). All of them will be alive while processed (since the parent will be) but this sounds pretty obscure.

Comment (Deexie) [reply api/task_manager.cc:211]: Which one do you find better?

**Comment by Deexie** [comment]
The PR with changes to api without recursive status:  #12338

**Comment by bhalevy** [comment]
please turn into draft in the meanwhile

**Comment by Deexie** [comment]
- deleted commits that were moved to aforementioned PRs
- stayed with queue approach; copying each foreign ptr

**Comment by Deexie** [comment]
Stays in draft until #12338 gets merged

**Comment by Deexie** [comment]
- rebased on master

**Comment by bhalevy** [comment]
@Deexie please update the PR title (it becomes part of the short log message of the merge commit)

File: api/task_manager.cc [hunk 6]
```diff
@@ -179,7 +180,51 @@ void set_task_manager(http_context& ctx, routes& r) {
                 return make_foreign(task);
             });
         }));
-        co_return co_await retrieve_status(std::move(task));
+        auto s = co_await retrieve_status(task);
+        co_return make_status(s);
```
Comment (bhalevy) [inline api/task_manager.cc:184]: Please extract this transition into a separate patch, before adding `get_task_status_recursively`

File: api/task_manager.cc [hunk 7]
```diff
@@ -179,7 +180,51 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+    });
+
+    tm::get_task_status_recursively.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        using chunked_statuses = utils::chunked_vector<full_task_status>;
+
+        auto& _ctx = ctx;
+        auto id = tasks::task_id{utils::UUID{req->param["task_id"]}};
+        std::queue<tasks::task_manager::foreign_task_ptr> q;
+        chunked_statuses res;
```
Comment (bhalevy) [inline api/task_manager.cc:193]: If `chunked_statuses` is used only here, there's no point in a typedef.

File: api/task_manager.cc [hunk 8]
```diff
@@ -179,7 +180,51 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+
+        // Get requested task.
+        auto task = co_await tasks::task_manager::invoke_on_task(_ctx.tm, id, std::function([] (tasks::task_manager::task_ptr task) -> future<tasks::task_manager::foreign_task_ptr> {
+            auto state = task->get_status().state;
+            if (state == tasks::task_manager::task_state::done || state == tasks::task_manager::task_state::failed) {
+                task->unregister_task();
+            }
+            co_return std::move(task);
```
Comment (bhalevy) [inline api/task_manager.cc:201]: copy elision would take care of the move, so `co_return task` should be just as good.

File: api/task_manager.cc [hunk 9]
```diff
@@ -179,7 +180,51 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+
+        // Push children's statuses in BFS order.
+        q.push(co_await task.copy());   // Task cannot be moved since we need it to be alive during whole loop execution.
+        while (!q.empty()) {
+            auto& current = q.front();
+            res.push_back(co_await retrieve_status(current));
+            for (auto& t : current->get_children()) {
+                q.push(co_await t.copy());
+            }
+            q.pop();
```
Comment (bhalevy) [inline api/task_manager.cc:212]: nice!

File: api/task_manager.cc [hunk 10]
```diff
@@ -179,7 +180,51 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+            co_await s.write("[");
+            std::string delim = "";
+            for (auto& status: res) {
+                co_await s.write(std::exchange(delim, ", "));
+                co_await formatter::write(s, make_status(status));
+            }
+            co_await s.write("]");
+            co_await s.close();
+        };
+        co_return std::move(f);
```
Comment (bhalevy) [inline api/task_manager.cc:227]: do we have to move f? wouldn't copy elision do that for us?

**Comment by Deexie** [comment]
- moved retrieve_status change to separate patch
- deleted chunked_statuses typedef
- deleted unnecessary move's

**Review by bhalevy — APPROVED** [review]

File: api/task_manager.cc [hunk 11]
```diff
@@ -183,6 +183,47 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+        auto id = tasks::task_id{utils::UUID{req->param["task_id"]}};
+        std::queue<tasks::task_manager::foreign_task_ptr> q;
+        utils::chunked_vector<full_task_status> res;
+
+        // Get requested task.
+        auto task = co_await tasks::task_manager::invoke_on_task(_ctx.tm, id, std::function([] (tasks::task_manager::task_ptr task) -> future<tasks::task_manager::foreign_task_ptr> {
+            auto state = task->get_status().state;
+            if (state == tasks::task_manager::task_state::done || state == tasks::task_manager::task_state::failed) {
+                task->unregister_task();
+            }
```
Comment (denesb) [inline api/task_manager.cc:198]: Why is a query for the task status has to take care of unregistering the task?

File: api/task_manager.cc [hunk 12]
```diff
@@ -183,6 +183,47 @@ void set_task_manager(http_context& ctx, routes& r) {
...
+            co_return task;
+        }));
+
+        // Push children's statuses in BFS order.
+        q.push(co_await task.copy());   // Task cannot be moved since we need it to be alive during whole loop execution.
+        while (!q.empty()) {
+            auto& current = q.front();
+            res.push_back(co_await retrieve_status(current));
+            for (auto& t : current->get_children()) {
```
Comment (denesb) [inline api/task_manager.cc:207]: Can this be called from any shard? Won't this cause corruption if the children list is concurrently modified on the task's home shard?

**Review by denesb — CHANGES_REQUESTED** [review]

File: api/task_manager.cc (hunk 11 above)
Comment (Deexie) [reply api/task_manager.cc:198]: Tasks are unregistered when they are finished (either successfully or not) and one of the following occurred:
- ttl time passed since the task has finished
- the task status was retrieved with an api.

Thus, it seemed reasonable to do it openly in the api call (as it is done for `get_task_status` and `wait_task`).
Maybe `task::get_status_and_unregister()`  should rather do that and be called from here?

Comment (Deexie) [reply api/task_manager.cc:198]: Also, I was wondering how `get_task_status_recursively` should behave regarding unregistration of tasks (unregister all of them/none/base task only?). I feel like it shouldn't do less than `get_task_status` as it is somehow its extension, but unregistering the whole tree seems to be a little to much (many calls to other shards). Do you agree?

**Comment by Deexie** [comment]
- changed foreach on `_children` to for

File: api/task_manager.cc (hunk 11 above)
Comment (denesb) [reply api/task_manager.cc:198]: Sorry, I somehow missed your response.

I think it is odd that we remove the task after querying its status. I think we should always let it live until the TTL expires. But that is out of the scope of this PR.

Comment (denesb) [reply api/task_manager.cc:198]: Won't unregistering the top level task also get rid of the other ones? Or they become orphaned until TTl expires?

Comment (Deexie) [reply api/task_manager.cc:198]: They are still stored in module until TTL expires or their status is requested.

Comment (denesb) [reply api/task_manager.cc:198]: Ok, I think it is fine as-is for now, but as a follow-up I think we should consider removing this unregistering on status query. There might be more than one operators wanting to have a look at the outcome of a task, or one might want to look at it again. I think the current practice is confusing. @bhalevy what do you think?

**Review by denesb — APPROVED** [review]
